Parameter passing prototype

jennuine opened the pull request:
This is the first PR for a series of PRs (i.e., incremental PRs with new features will be merged into this one). Everything will be merged into `sdf10` once the whole series is completed.

Parameter passing reference: http://sdformat.org/tutorials?tut=param_passing_proposal&branch=param_passing

jennuine commented:
30b2317: Parser checks for `//include/experimental:params` tag and added test

azeey commented on test/integration/include_custom_model.sdf:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" ?>
+<sdf version="1.7">
```
To make this a valid xml, I think we need to define the `experimental` namespace. My suggestion would be to use `http://sdformat.org/schemas/experimental`.
```suggestion
<sdf version="1.7" xmlns:experimental="http://sdformat.org/schemas/experimental">
```

jennuine replied:
Done 5c167b32c5d79d57292702ebfb4114b39e4506cb

azeey commented on test/integration/param_passing.cc:
```diff
@@ -0,0 +1,68 @@
+{
+  const std::string modelRootPath =
+    sdf::filesystem::append(PROJECT_SOURCE_PATH, "test", "integration",
+                            "model") + "/";
```
Would this cause issues in windows? I believe `sdf::filesystem::append` uses `\` in windows.

jennuine replied:
5c167b32c5d79d57292702ebfb4114b39e4506cb

azeey commented on test/integration/param_passing.cc:
```diff
@@ -0,0 +1,68 @@
+  sdf::setFindCallback(
+      [&](const std::string &_file)
+      {
+        return modelRootPath + _file;
```
Consider using `sdf::filesystem::append` here instead of `+`. This would also avoid the need for the `+ "/"` in `modelRootPath` above.

jennuine replied:
5c167b32c5d79d57292702ebfb4114b39e4506cb

azeey commented on test/integration/param_passing.cc:
```diff
@@ -0,0 +1,68 @@
+#include "sdf/Root.hh"
+#include "test_config.h"
+
+void print_errors(sdf::Errors &_errors)
```
nit: style `print_errors` -> `PrintErrors`

jennuine replied:
5c167b32c5d79d57292702ebfb4114b39e4506cb

iche033 commented on test/integration/model/nested_include/base/base.sdf:
```diff
@@ -0,0 +1,168 @@
+      </visual> -->
+
+      <sensor name="camera" type="camera">
+      <!-- <sensor name="camera" type="thermal"> -->
```
remove this comment?

jennuine replied:
a9238215ab6ecae86f1d5b93aa72e75fdfec977b

iche033 commented on test/integration/model/nested_include/base/base.sdf:
```diff
@@ -0,0 +1,168 @@
+            <size>0.1 0.1 0.1</size>
+          </box>
+        </geometry>
+      </visual> -->
```
is this still needed? I think this block gets added later by `add` action?

iche033 replied:
there is also the same commented out element in `robot/model.sdf` as well

jennuine replied:
a9238215ab6ecae86f1d5b93aa72e75fdfec977b

iche033 commented on test/integration/model/nested_include/top_camera/top_camera.sdf:
```diff
@@ -0,0 +1,46 @@
+          <sphere>
+            <radius>0.05</radius>
+          </sphere>
+        </geometry> -->
```
this shows the new <geometry> block that is going to be the replacement of the current one? maybe add a comment?

jennuine replied:
a9238215ab6ecae86f1d5b93aa72e75fdfec977b

iche033 commented on test/integration/model/nested_include/top_camera/top_camera.sdf:
```diff
@@ -0,0 +1,46 @@
+      </visual>
+
+      <sensor name="camera" type="camera">
+        <!-- <pose>-0.2 0 0.3 0 0.0 3.14</pose> -->
```
this is the replacement element?

jennuine replied:
a9238215ab6ecae86f1d5b93aa72e75fdfec977b

iche033 reviewed (commented):
looks good. Just minor comments about either removing commented out blocks in the sdf or adding doc to describe what they are for

jennuine commented:
> Just minor comments about either removing commented out blocks in the sdf or adding doc to describe what they are for

I removed the unneeded comments in a9238215ab6ecae86f1d5b93aa72e75fdfec977b and also added a test that compares the result with the expected output for nested models

azeey commented on src/ParamPassing.cc:
```diff
@@ -0,0 +1,675 @@
+      // need to add `|| elemName == "model"` since nested models are allowed
+      if (elemName == "link")
+      {
+        useParentDesc = true;
```
I'm not quite sure if I understand why `link` is being treated differently here. Is the parent element assumed to be `model`? If so should we include `joint` and `frame` as well?.

azeey replied:
per VC, we don't need this `if` block. Everything in the `else` block can be hoisted up.

jennuine replied:
3b28f1faece56bd1533ff32a4cb4b85536dded65

azeey commented on test/integration/param_passing.cc:
```diff
@@ -0,0 +1,175 @@
+TEST(ParamPassingTest, ExperimentalParamsTag)
+{
+  const std::string modelRootPath =
+    sdf::filesystem::append(PROJECT_SOURCE_PATH, "test", "integration",
```
We have recently been using `sdf::testing::TestFile` for creating these type of paths. Can you update all of the paths to use that?

jennuine replied:
e8919a107af56891d1acf75ec61bcd646fdf0a7c

azeey commented on test/integration/model/nested_include/base/base.sdf:
```diff
@@ -0,0 +1,158 @@
+    </link>
+
+    <include>
+      <uri>top</uri>
```
Can you set the name to `top_model` here just to make it clear that the model from uri `top` is called `top_model`.

jennuine replied:
e8919a107af56891d1acf75ec61bcd646fdf0a7c

azeey commented on test/integration/model/robot/model.sdf:
```diff
@@ -0,0 +1,324 @@
+      <child>caster</child>
+    </joint>
+
+    <link name="nested_links">
+      <pose>0 0 0 0 0 0</pose>
+      <link name="link1">
+        <link name="link2"/>
+      </link>
+    </link>
```
This is not valid SDFormat because a link cannot be nested under another link. Is it being used for testing error conditions? If so, can you add a comment stating so.

jennuine replied:
For some reason I thought nested links were possible, I modified the file to be a nested model instead 3b28f1faece56bd1533ff32a4cb4b85536dded65

jennuine commented on src/ParamPassing.hh:
```diff
@@ -0,0 +1,147 @@
+    /// \brief Replace an element with another element
+    /// \param[in] _newElem The replacement element
+    /// \param[out] _origElem The element to be replaced
+    void replace(const ElementPtr _newElem, ElementPtr _origElem);
```
`replace` doesn't emit errors since the errors differ when the `action` is defined in the `element_id` or if defined in the children elements. 

This is the error that occurs when the `action` is defined with `element_id`: 
https://github.com/osrf/sdformat/blob/81112de2c5fc18a071e8eacb9b9b2b2d3cfb9fae/src/ParamPassing.cc#L186-L195

These are the errors when the `action` is defined in the children of `element_id` (since there could be several children with different defined `action`s, this gave a clearer error for which child caused the issue and was skipped):
https://github.com/osrf/sdformat/blob/81112de2c5fc18a071e8eacb9b9b2b2d3cfb9fae/src/ParamPassing.cc#L479-L503

jennuine commented on src/ParamPassing.cc:
```diff
@@ -0,0 +1,675 @@
+    if (!_elem->HasAttribute(attrName))
+    {
+      // TODO(jenn) add attributes for namespaced elements?
+      if (attrName.find(":") != std::string::npos)
```
This is needed because I'm reading the attributes from tinyxml (listed in `<experimental:params>`) and adding the attribute to the original element if there isn't one or it's a custom attribute (not using functions from parser.cc). For example, if a user wanted to add an attribute `test` this would fail, but if they used `custom:test` it would succeed. Should I change it to skip the `_elem->HasAttribute` check and add the attribute(s) anyway?

jennuine commented on src/ParamPassing.cc:
```diff
@@ -0,0 +1,675 @@
+    }
+    else if (actionStr == "replace")
+    {
+      ElementPtr newElem = initElementDescription(childElemXml, _errors);
+      if (!newElem)
+        continue;
+
+      if (!readXml(childElemXml, newElem, _errors))
+      {
+        _errors.push_back({ErrorCode::ELEMENT_INVALID,
+          "Unable to convert XML to SDF. Skipping element replacement:\n"
+          + ElementToString(childElemXml)
+        });
+        continue;
+      }
```
See https://github.com/osrf/sdformat/pull/413#discussion_r655759787

jennuine commented on src/ParamPassing.cc:
```diff
@@ -0,0 +1,675 @@
+
+  if (!initFile(filename, elemDesc))
+  {
+    // TODO(jenn) not sure if we should load the element anyway
```
No, the first code snippet would reach this error:
https://github.com/osrf/sdformat/blob/81112de2c5fc18a071e8eacb9b9b2b2d3cfb9fae/src/ParamPassing.cc#L444-L451

Yes, the 2nd snippet is not possible yet but this is actually possible:
```
<visual element_id="v1" name="visual" action="add">
   <foo:custom_elem/>
</visual>
```

I'll make sure to add that example to the tutorial

jennuine replied:
Hows this? https://github.com/osrf/sdf_tutorials/pull/51/commits/edff4f197d4e244ed2eed9c5126979f105c1b648

azeey replied:
Thanks. I've made a suggestion there.

jennuine commented on src/ParamPassing.cc:
```diff
@@ -0,0 +1,675 @@
+      elemChild = getElementByName(_elem, xmlChild);
+      if (elemChild == nullptr)
+      {
+        const tinyxml2::XMLElement *xmlParent = _xml->Parent()->ToElement();
```
Yes, the 2 places where `remove` is called is:

(1) when the `action` is in the `element_id`: https://github.com/osrf/sdformat/blob/81112de2c5fc18a071e8eacb9b9b2b2d3cfb9fae/src/ParamPassing.cc#L178

(2) when the `action` is in the children of `element_id`:
https://github.com/osrf/sdformat/blob/81112de2c5fc18a071e8eacb9b9b2b2d3cfb9fae/src/ParamPassing.cc#L402

jennuine commented on src/parser.cc:
```diff
@@ -890,9 +891,11 @@ bool readXml(tinyxml2::XMLElement *_xml, ElementPtr _sdf, Errors &_errors)
 
-    if (i == _sdf->GetAttributeCount())
+    std::string attrName = attribute->Name();
+    if (i == _sdf->GetAttributeCount()
```
It does work, thanks for the tip https://github.com/osrf/sdf_tutorials/pull/51/commits/edff4f197d4e244ed2eed9c5126979f105c1b648

jennuine commented on src/ParamPassing.hh:
```diff
@@ -0,0 +1,147 @@
+    // parse and update data into
+    /// \param[out] _errors Captures errors found during parsing
+    void updateParams(tinyxml2::XMLElement *_childXmlParams,
+                      SDFPtr _includeSDF, Errors &_errors);
```
No reason, fixed d7aec4767ff0847c04de84e4ace11c435926d5d8

jennuine commented on src/ParamPassing.hh:
```diff
@@ -0,0 +1,147 @@
+    /// the add action
+    /// \return ElementPtr to the specified element, nullptr if element could
+    /// not found
+    ElementPtr getElementById(const SDFPtr _sdf,
```
d7aec4767ff0847c04de84e4ace11c435926d5d8

jennuine commented on src/ParamPassing.cc:
```diff
@@ -0,0 +1,675 @@
+      }
+    }
+
+    if (!xmlChild->NoChildren())
```
e8919a107af56891d1acf75ec61bcd646fdf0a7c

azeey reviewed (approved):
Great work!! And thank you for iterating.
